File: JavaToKotlin/app/src/main/java/io/dustin/apps/board/api/usecase/blockeduser/WriteBlockedUserUseCase.java
package io.dustin.apps.board.api.usecase.blockeduser;

import io.dustin.apps.board.domain.blockeduser.model.BlockedUser;
import io.dustin.apps.board.domain.blockeduser.model.dto.BlockedUserDto;
import io.dustin.apps.board.domain.blockeduser.service.WriteBlockedUserService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class WriteBlockedUserUseCase {

    private final WriteBlockedUserService writeBlockedUserService;

    /**
     유저를 차단하면
     1. 차단한 유저의 게시물이 보이면 안된다 -- > 이건 querydsl로 해결한다.
     2. 만약 서로 팔로우 상태중이라면(맞팔) 끊어져야 한다 -- > 유스케이스에 추가해야할듯
     * */

    /** 유저와 차단한 상대 팔로우(팔로잉) 취소(db 삭제) 로직 필요 */
    /** 유저와 차단한 상대 팔로우(팔로잉) 수 감소 로직 필요 */

    public BlockedUserDto execute(Long fromUserId, Long toUserId) {
        BlockedUser blockedUser = writeBlockedUserService.create(fromUserId, toUserId);
        BlockedUserDto dto = BlockedUserDto.from(blockedUser);
        return BlockedUserDto.from(blockedUser);
    }
}

File: apps/src/main/kotlin/io/dustin/apps/board/api/usecase/blockeduser/WriteBlockedUserUseCase.kt
package io.dustin.apps.board.api.usecase.blockeduser

import io.dustin.apps.board.domain.blockeduser.model.dto.BlockedUserDto
import io.dustin.apps.board.domain.blockeduser.service.ReadBlockedUserService
import io.dustin.apps.board.domain.blockeduser.service.WriteBlockedUserService
import io.dustin.apps.common.exception.DuplicatedException
import io.dustin.apps.common.utils.duplicate
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class WriteBlockedUserUseCase (
    private val readBlockedUserService: ReadBlockedUserService,
    private val writeBlockedUserService: WriteBlockedUserService,
) {
    @Transactional
    fun execute(fromUserId: Long, toUserId: Long): BlockedUserDto {
        val blockedOjb = readBlockedUserService.getBlockedObj(fromUserId,toUserId)

        if(blockedOjb == null){
            val blockedUser = writeBlockedUserService.create(fromUserId, toUserId)
            return BlockedUserDto.from(blockedUser)
        }
        else{
            return throw duplicate("이미 차단한 유저입니다.")
        }
    }
}
